### ISSUE TITLE
Single step issues in various simulators

### ISSUE THREAD

**Issue by archivisth** [simh/simh#388]
Rich Cornwell pointed out that the SimH 'single step' model is based on the assumption of one simulated timer tick = one instruction. This won't work for simulators that time by memory cycle, and so on. I provided Rich with an alternative way of implementing stepping for those simulators.

However, his comment made me realize that some of the existing simulators also violate the 'one tick = one instruction' assumption. I've only looked at the 3.X code base, so I can't comment on simulators that run only in the 4.X context.

PDP8 - the FPP8, if running, will charge a clock tick for each instruction that it executes. Because the FPP8 is modeled as an IO device, it may cause the stepping model to get out of sync. I don't think this is worth fixing; the FPP8 is rarely used.

PDP11 - the CIS instructions charge for each major iteration of their loops but only test for an interrupt every 100 cycles or so. The instruction will stop with FPD set and state in the registers/on the stack, just like a real PDP11. Again, I don't it's worth doing anything about this.

VAX - the character string instructions, except MATCHC, will count down for each major iteration but do not test for interrupts or stop in mid instruction. This makes STEP N (N > 1) across a character string instruction unpredictable. MATCHC will count down for each substring comparison and will stop in mid instruction if there is an interrupt or event. (On a real 780, the worst-case MATCH ran for half an hour.) The instruction is packed up and FPD is set. Thus, you can STEP through a MATCHC instruction one substring compare at a time. Other iterative instructions (CRC, decimal, POLY) run to completion; they never set FPD. I don't think it's worth doing anything about stepping in the VAX.

PDP10 - because interrupts are the only way out of infinite indirect and XCT loops, the PDP10 counts down the event clock for each memory reference beyond "the basics", including the loops in BLT/BLTU, and interrupts out of these situations.  (The extended instructions always run to completion.) As a result, STEP really doesn't work if the instruction being stepped over has multi-level indirects or XCTs or is a BLT/BLTU. The event test gets a step-count-expired event and returns to the simulator, and execution will start over with the same instruction.

In summary: I propose to document but not otherwise deal with the STEP anomalies in the PDP8, PDP11, and VAX. However, I believe the PDP10 needs to be modified along the lines I proposed; it really is using memory cycles more than instructions for its timing, at least in the main CPU.

**Comment by markpizz** [comment]
I actually saw this recently in the VAX780 simulator while working on the new timing code.

What I saw was that sim_interval was decremented once for each byte processed in the string instructions.  Conceptually that is reasonable, but I observed that when it goes negative nothing is done until after the instruction completes.  Given the fact that these instructions could manipulate very large strings (32bit length field), it could go VERY negative.  The net effect is that the only time the adjusted value of sim_interval will be observed for event dispatching DURING the execution of the instruction would be if the instruction's execution were to be suspended to to a page fault as the memory being manipulated crossed page boundaries.  I didn't want to explore what it would take to cleanly suspend execution to process events (while still being able to reliably exit back out of sim_instr() if any event's execution returned other than SCPE_OK).

What I did observe while I was looking at this, was that there is code early in the VMS boot for this machine, which uses a MOVC3 or MOVC5 instruction to touch a very large amount of memory,  The consequence of this was that these per byte decrements of sim_interval was causing the observed instruction execution rate which started at about 14-15 MIPS on my desktop to jump up to 160MIPS.  This skewed the clock calibration for a while which at the time I believed was affecting some of the results I was trying to achieve.  I came up with an adjustment strategy to sim_interval which allows it to be more aligned with the fact that, for the most part, sim_interval counts instructions.  This change is implemented in 15550563961568e2e6773072dc0bf44b0978fbd1

In the end, this is unrelated to the behavior of STEP (at least for a single instruction step) since its effects are observed/addressed at the start of each instruction.  So, I agree that otherwise nothing special should be done on the VAX to address stepping.

**Comment by hrlzm** [comment]
Comment about the PDP10:  I can't say if the extend instructions don't accept interrupts mid-instruction,
they SHOULD, but they certainly can take a page fault and get restarted.

**Comment by R-Voorhorst** [comment]
From the processor reference manual regarding the string instructions page 2-98 :

If an interrupt or page failure occurs during execution of a string move
or compare, the accumulators are adjusted for what has already been done.
Afterwards the instruction resumes as though starting at the beginning,
but manipulates substrings that are simply those parts of the original
strings left from where the instruction was interrupted.

So in compliance with the general Pdp10 architecture interrupts etc. are taken.

**Comment by archivisth** [comment]
VAX strings are limited to 64KB - 16b lengths. So yes, the counter can 
go negative, but it can't wrap.

**Comment by markpizz** [comment]
I stand corrected.

I know I remember someone once saying that you could zero all of memory with a single instruction.  I never tried it, so clearly they were wrong.

Meanwhile, that VAX boot activity which crazily skewed the instruction execution rate must have been doing a lot of MOVC instructions with pretty decent sizes to have skewed the rate by more than 10X.
